src/include/storage/storage_defs.h

```diff
@@ -416,30 +416,43 @@ class VarlenEntry {
    */
   template <bool EqualityCheck>
   static bool CompareEqualOrNot(const VarlenEntry &left, const VarlenEntry &right) {
-    TERRIER_ASSERT(left.Size() >= 0, "Left VarlenEntry has negative size?");
-    TERRIER_ASSERT(right.Size() >= 0, "Right VarlenEntry has negative size?");
-
     // Compare the size and prefix in one fell swoop, ignoring the sign bit indicating reclaimability.
-    uint8_t left_first = *reinterpret_cast<const uint8_t *>(&left);
-    uint8_t right_first = *reinterpret_cast<const uint8_t *>(&right);
-    bool first_byte_same = (left_first << 1) == (right_first << 1);
-    bool following_bytes_same =
-        0 == std::memcmp(reinterpret_cast<const char *>(&left) + 1, reinterpret_cast<const char *>(&right) + 1,
-                         sizeof(left.size_) + PrefixSize() - 1);
-    if (first_byte_same && following_bytes_same) {
-      // Prefix and length are equal.
+    const uint64_t left_size_prefix = *reinterpret_cast<const uint64_t *>(&left);
+    const uint64_t right_size_prefix = *reinterpret_cast<const uint64_t *>(&right);
+    // Mask off the reclaimability bit and compare.
+    // Note that due to endianness, when reading this out as a uint64_t the reclaim bit is NOT the sign bit.
+    constexpr uint64_t remove_reclaim_mask = 0xffffffff7fffffff;  // You're just gonna have to believe me
```

Comment:
This seems incredible unintuitive - you got it wrong in an earlier commit.

If we're going to stick with this, I'd like to see a performance benchmark against a simpler version that checks the 7 bytes after the first as a "lossy" filter, then validate using the actual size. I think this is easier to understand and will probably have equivalent performance without the need for magic numbers.

Reply:
Per discussion in meeting, leaning towards checking size followed by prefix separately. It is not possible to check the 7 bytes together.